### DISCUSSION TITLE
Azure Authentication

### DISCUSSION THREAD

**Discussion by yesoreyeram** [grafana/grafana-infinity-datasource#404] [Specific APIs] [answered]
# Steps to connect to Azure APIs.

1. Create a new service client in [azure portal](https://portal.azure.com/#view/Microsoft_AAD_IAM/ActiveDirectoryMenuBlade/~/RegisteredApps)
2. Create a client secret for the client created above
3. Note down the Client ID, Client Secret and Tenant ID
4. Give reader/monitoring reader access to the resources/subscriptions as necessary
5. Install the infinity plugin in Grafana and add data source for the same
6. Expand Authentication section and select "OAuth2"
7. Select "Cllient Credentials" as OAuth2 type
8. Specify the Client ID
9. Specify the Client Secret
10. Specity the Token URL `https://login.microsoftonline.com/<TENENT_ID>/oauth2/token`. Replace <TENENT_ID> with yours
11. Leave the Scopes section empty
12. Add the following Endpoint param
     1. Key : `resource` Value: `https://management.azure.com/`
13. If you are using Infinity 1.0.0+, then specify `https://management.azure.com/` as an allowed URL.
14. Click Save and Test.
15. Click the `Explore` button
16. Configure the query
   1. Specify `json` as query type
   2. Specify `url` as source
   3. Specify `default` parser ( only applicable for Infinity 1.0.0+)
   4. Specify `table` as format
   5. Method : GET
   6. URL  : https://management.azure.com/subscriptions?api-version=2020-01-01
17. Click Run Query and you will see the results

## Config Editor
<img width="1190" alt="image" src="https://user-images.githubusercontent.com/153843/190214740-c8b548f9-ef64-4399-941f-41df5f631fdc.png">

## Query using Default Parser

<img width="1791" alt="image" src="https://user-images.githubusercontent.com/153843/190215987-d21424dd-f12e-4ea3-be79-f2959e960dc0.png">

## Query using Backend Parser

<img width="1792" alt="image" src="https://user-images.githubusercontent.com/153843/190216309-12a8f8f7-3a9d-4b90-bc48-0919b0a94e72.png">

## Query using UQL Parser

<img width="1088" alt="image" src="https://user-images.githubusercontent.com/153843/190216710-d39d779c-d984-4fe8-b450-f55ebfcd6496.png">

**Comment by itsjusth** [comment, marked as answer]
Thank you! This worked perfectly.

**Comment by switchgl** [comment]
Thank you for the above, this works as expected!

What would the configuration be if I'm trying to access data in a blob storage container? I understand the resource setup would be https://storage.azure.com/ which I've set along with adding it as an allowed host however I get an error "Cannot read properties of null (reading 'trim')" using an older version of the infinity plugin and error "400 Authentication information is not given in the correct format. Check the value of Authorization header." using the latest version of the infinity plugin.

is there a known working configuration setup for accessing azure blob storage containers using AD auth? essentially I have a csv file stored in an azure blob storage container that I want to display in a dashboard.

Using SAS token authentication works fine however I have only managed to get it working by pasting the full URL into the URL field in the dashboard itself which has the SAS token visible (not something I want visible) and I can't seem to get it working setting the SAS token in the data source area itself as it doesn't seem to parse it properly so errors no matter what settings I choose.

**Reply by yesoreyeram** [reply]
Hiya.. sorry for the delay. I don't think azure blob storage supports this type of authentication. I am planning to add support for providing azure blob storage key authentication in an upcoming version. Thanks for trying.

**Reply by yesoreyeram** [reply]
version2.1.0 supports reading content from azure blob storage.

**Reply by crlgb** [reply]
Do you intend to support Azure Table Storage in future versions as well? :)

**Reply by pltzr** [reply]
It also seems possible to connect to an Azure Storage Account with OAuth as well.

Follow the steps above but for Step 12, use `https://storage.azure.com/` and in the URL, Headers & Params section add a header `x-ms-version` of `2017-11-09`.

Of course, the application you are using needs to have at least `Storage Blob Data Reader` access to the blob you're trying to access.

**Comment by paige-muon** [comment]
@yesoreyeram just a heads up, this type of authentication is broken in the latest version of grafana: https://github.com/yesoreyeram/grafana-infinity-datasource/issues/668
I opened a support request with grafana and they confirmed it is something in the plugin that they aren't able to do anything about.

**Reply by yesoreyeram** [reply]
Does this works in older version of grafana? BTW, You meant this is grafana issue or plugin issue? 

If it works in older version of grafana, i suggest to use that as workaround to configure. Then you can update the grafana. 

I can look into this in detail after my parental leave. Thanks.

**Reply by paige-muon** [reply]
Oauth2 authentication in this plugin broke when grafana was updated, but I followed up with grafana and they are not able to do anything about it. I hoped they would have seen other impacts from whatever causes this elsewhere in grafana and then fix it, but apparently not.

**Reply by yesoreyeram** [reply]
Hi @paige-muon - this is already fixed in latest version. Can you check and confirm?

**Reply by paige-muon** [reply]
Yup, this was an old comment. Everything is working now

**Comment by ganesh1255** [comment]
The steps mentioned above worked fine with plugin version 2.9.6, but with the latest version of the plugin the error message shows
oauth2: "invalid request": The 'resource' request parameter is not supported.

**Reply by yesoreyeram** [reply]
Can you help with reproduction steps? If reproducible, create a bug report please.

**Comment by Wesztman** [comment]
@yesoreyeram It would be awsome if you could add to this guide a little more info on step 4, i'm stuck with 

`error received: unsuccessful HTTP response. 401 Unauthorized`

And I don't understand what I need to do to actually give Grafana/Infinity access to pass through the authentication on my web app :(  I need to read out my /health endpoint from my azure web app with MS authentication configured.

I've done everythin with client id and secrets etc

Cheers
Carl

**Comment by pabes74** [comment]
@yesoreyeram  I get a "error while performing the infinity query. unsuccessful HTTP response code status code : 401 Unauthorized" when exploring data using the infinity datasource.  Here are the screenshots. 
![image](https://github.com/user-attachments/assets/e38128ed-e08f-4e9d-8c3d-d5978ebcbf9c)

![image](https://github.com/user-attachments/assets/30e1e387-4807-4de4-bc92-6b7a66c1e7e4)

I tested my ClientID and ClientSecret with a the following curl :` curl --location 'https://login.microsoftonline.com/<tenantid>/oauth2/token' --form 'client_id="<clientid>"' --form 'client_secret="<clientsecret>"' --form 'resource="https://management.azure.com/"' --form 'grant_type="client_credentials"'`. This works fine and returns a working token. So it looks like my credentials are valid. I'm iu doing something wrong? Any help would be appreciated.

**Reply by Wesztman** [reply]
@pabes74 , I know I solved this.. But I accidentally removed my data source 😅 and I realize I didn't write down in my own comment above how I solved this... Did you solve it? Do you remember how?
